services/apps/data_sink_worker/src/service/activity.service.ts

```diff
@@ -1174,7 +1174,31 @@ export default class ActivityService extends LoggerBase {
 
     this.log.trace(`[ACTIVITY] We have ${preparedActivities.length} intermediate results!`)
 
-    const preparedForUpsert = preparedActivities.filter((a) => a.payload)
+    const activitiesWithPayload = preparedActivities.filter((a) => a.payload)
+
+    const uniqueConstraintKeys = new Set<string>()
+    const preparedForUpsert = []
+
+    for (const item of activitiesWithPayload) {
+      // only include non-null, non-empty values in the key
+      const key = [
+        item.payload.timestamp,
+        item.payload.platform,
+        item.payload.type,
+        item.payload.sourceId,
+        item.payload.channel,
+        item.payload.segmentId,
+      ]
+        .filter((v) => v !== undefined && v !== null && v !== '')
```

Comment:
To avoid any weird undefined behavior I would instead of ignoring empty values replace them with `<empty>` string or smt like that just to keep the format of a key consistent.